File: 5.longest-palindromic-substring.java
/*
 * @lc app=leetcode id=5 lang=java
 *
 * [5] Longest Palindromic Substring
 */

// @lc code=start
class Solution {
    public String longestPalindrome(String s) {
        if (s == null || s.length() == 0) {
            return "";
        }
        int start = 0,end =0,len=0;
        
        for (int i = 0; i < s.length(); i++) {
            int len1 = isPalindrome(s, i, i);
            int len2 = isPalindrome(s, i, i + 1);
            len = Math.max(len1, len2);
            if(len > end - start+1) {
                start = i - (len - 1) / 2;
                end = i + len / 2;
            }
        }
        return s.substring(start, end + 1);
    }
    
    private int isPalindrome(String s, int left, int right) {   
        int L = left, R = right;
        while (L >= 0 && R < s.length() && s.charAt(L) == s.charAt(R)) {
            L--;
            R++;
        }
        return R - L - 1;
    }
}
// @lc code=end

File: 5.longest-palindromic-substring.py
#
# @lc app=leetcode id=5 lang=python3
#
# [5] Longest Palindromic Substring
#

# @lc code=start
class Solution:
    def longestPalindrome(self, s: str) -> str:
        def expandAroundCenter(left:int, right:int)->str:
            while left >= 0 and right < len(s) and s[left] == s[right]:
                left -= 1
                right += 1
            return s[left + 1:right]
        
        start,end = 0,0
        for i in range(len(s)):
            len1 = len(expandAroundCenter(i,i))
            len2 = len(expandAroundCenter(i,i+1))
            maxLen = max(len1,len2)
            if maxLen > end - start:
                start = i - (maxLen-1)//2
                end = i + (maxLen)//2
        return s[start:end+1]
        
        
        
# @lc code=end

